File: app/src/main/java/io/neoterm/view/eks/ExtraButton.java
package io.neoterm.view.eks;

import android.view.KeyEvent;
import android.view.View;

import io.neoterm.R;
import io.neoterm.backend.TerminalSession;
import io.neoterm.view.TerminalView;

/**
 * @author kiva
 */

public abstract class ExtraButton implements View.OnClickListener {
    public static final String KEY_ESC = "Esc";
    public static final String KEY_TAB = "Tab";
    public static final String KEY_CTRL = "Ctrl";
    public static final String KEY_PAGE_UP = "PgUp";
    public static final String KEY_PAGE_DOWN = "PgDn";
    public static final String KEY_HOME = "Home";
    public static final String KEY_END = "End";
    public static final String KEY_ARROW_UP = "▲";
    public static final String KEY_ARROW_DOWN = "▼";
    public static final String KEY_ARROW_LEFT = "◀";
    public static final String KEY_ARROW_RIGHT = "▶";

    public String buttonText;

    @Override
    public abstract void onClick(View view);

    public static void sendKey(View view, String keyName) {
        int keyCode = 0;
        String chars = null;
        switch (keyName) {
            case KEY_ESC:
                keyCode = KeyEvent.KEYCODE_ESCAPE;
                break;
            case KEY_TAB:
                keyCode = KeyEvent.KEYCODE_TAB;
                break;
            case KEY_ARROW_UP:
                keyCode = KeyEvent.KEYCODE_DPAD_UP;
                break;
            case KEY_ARROW_LEFT:
                keyCode = KeyEvent.KEYCODE_DPAD_LEFT;
                break;
            case KEY_ARROW_RIGHT:
                keyCode = KeyEvent.KEYCODE_DPAD_RIGHT;
                break;
            case KEY_ARROW_DOWN:
                keyCode = KeyEvent.KEYCODE_DPAD_DOWN;
                break;
            case KEY_PAGE_UP:
                keyCode = KeyEvent.KEYCODE_PAGE_UP;
                break;
            case KEY_PAGE_DOWN:
                keyCode = KeyEvent.KEYCODE_PAGE_DOWN;
                break;
            case KEY_HOME:
                keyCode = KeyEvent.KEYCODE_MOVE_HOME;
                break;
            case KEY_END:
                keyCode = KeyEvent.KEYCODE_MOVE_END;
                break;
            case "―":
                chars = "-";
                break;
            default:
                chars = keyName;
        }

        if (keyCode > 0) {
            view.dispatchKeyEvent(new KeyEvent(KeyEvent.ACTION_DOWN, keyCode));
            view.dispatchKeyEvent(new KeyEvent(KeyEvent.ACTION_UP, keyCode));
        } else {
            TerminalView terminalView = (TerminalView) view.findViewById(R.id.terminal_view);
            TerminalSession session = terminalView.getCurrentSession();
            if (session != null) session.write(chars);
        }
    }
}

File: app/src/main/java/io/neoterm/view/eks/ExtraButton.kt
package io.neoterm.view.eks

import android.view.KeyEvent
import android.view.View

import io.neoterm.R
import io.neoterm.backend.TerminalSession
import io.neoterm.view.TerminalView

/**
 * @author kiva
 */

abstract class ExtraButton : View.OnClickListener {

    var buttonText: String? = null

    abstract override fun onClick(view: View)

    companion object {
        val KEY_ESC = "Esc"
        val KEY_TAB = "Tab"
        val KEY_CTRL = "Ctrl"
        val  KEY_ALT = "Alt"
        val KEY_PAGE_UP = "PgUp"
        val KEY_PAGE_DOWN = "PgDn"
        val KEY_HOME = "Home"
        val KEY_END = "End"
        val KEY_FN = "Fn"
        val KEY_BACK_SLASH = "\\"
        val KEY_ARROW_UP = "▲"
        val KEY_ARROW_DOWN = "▼"
        val KEY_ARROW_LEFT = "◀"
        val KEY_ARROW_RIGHT = "▶"
        val KEY_TOGGLE_FULL_SCREEN = "Fs"
        val KEY_TOGGLE_IME = "Im"

        fun sendKey(view: View, keyName: String) {
            var keyCode = 0
            var chars: String? = null
            when (keyName) {
                KEY_ESC -> keyCode = KeyEvent.KEYCODE_ESCAPE
                KEY_TAB -> keyCode = KeyEvent.KEYCODE_TAB
                KEY_ARROW_UP -> keyCode = KeyEvent.KEYCODE_DPAD_UP
                KEY_ARROW_LEFT -> keyCode = KeyEvent.KEYCODE_DPAD_LEFT
                KEY_ARROW_RIGHT -> keyCode = KeyEvent.KEYCODE_DPAD_RIGHT
                KEY_ARROW_DOWN -> keyCode = KeyEvent.KEYCODE_DPAD_DOWN
                KEY_PAGE_UP -> keyCode = KeyEvent.KEYCODE_PAGE_UP
                KEY_PAGE_DOWN -> keyCode = KeyEvent.KEYCODE_PAGE_DOWN
                KEY_HOME -> keyCode = KeyEvent.KEYCODE_MOVE_HOME
                KEY_END -> keyCode = KeyEvent.KEYCODE_MOVE_END
                "―" -> chars = "-"
                else -> chars = keyName
            }

            if (keyCode > 0) {
                view.dispatchKeyEvent(KeyEvent(KeyEvent.ACTION_DOWN, keyCode))
                view.dispatchKeyEvent(KeyEvent(KeyEvent.ACTION_UP, keyCode))
            } else {
                val terminalView = view.findViewById(R.id.terminal_view) as TerminalView
                val session = terminalView.currentSession
                session?.write(chars)
            }
        }
    }
}